fdbserver/DataDistribution.actor.cpp

```diff
@@ -1442,165 +1453,290 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
+			    .detail("AuditState", audit->coreState.toString());
+		}
+	} catch (Error& e) {
+		TraceEvent(SevInfo, "DDAuditStorageCoreError", self->ddId)
+		    .errorUnsuppressed(e)
+		    .detail("AuditID", (audit == nullptr ? UID() : audit->coreState.id))
+		    .detail("Range", req.range)
+		    .detail("AuditType", req.getType())
+		    .detail("RetryCount", (audit == nullptr ? 0 : audit->retryCount))
+		    .detail("Async", req.async);
+		if (e.code() == error_code_actor_cancelled) {
+			throw e;
+		} else if (audit == nullptr) {
+			throw retry();
+		} else if (audit->retryCount > SERVER_KNOBS->AUDIT_RETRY_COUNT_MAX) {
+			audit->coreState.setPhase(AuditPhase::Failed);
+			wait(persistAuditState(self->txnProcessor->context(), audit->coreState));
+			audit->actors.clear(true);
+			self->audits[audit->coreState.getType()].erase(audit->coreState.id);
+			if (!req.reply.isSet()) {
+				req.reply.send(audit->coreState.id);
+				TraceEvent(SevInfo, "DDAuditStorageCoreReplyFailed", self->ddId)
+				    .detail("AuditState", audit->coreState.toString())
+				    .detail("Async", req.async);
 			}
+		} else {
+			throw retry();
+		}
+	}
+	TraceEvent(SevDebug, "DDAuditStorageCoreEnd", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("AuditType", req.getType())
+	    .detail("Range", req.range)
+	    .detail("Async", req.async);
 
-			wait(audit->actors.getResult());
-			TraceEvent(SevInfo, "DDAuditStorageSuccess", self->ddId)
-			    .detail("AuditID", audit->id)
+	return Void();
+}
+
+ACTOR Future<Void> auditStorage(Reference<DataDistributor> self, TriggerAuditRequest req) {
+	loop {
+		try {
+			TraceEvent(SevDebug, "DDAuditStorageStart", self->ddId)
+			    .detail("AuditType", req.getType())
 			    .detail("Range", req.range)
-			    .detail("AuditType", req.type);
-			auditState.setPhase(AuditPhase::Complete);
-			wait(persistAuditState(self->txnProcessor->context(), auditState));
-			if (!req.async && !req.reply.isSet()) {
-				req.reply.send(audit->id);
-			}
+			    .detail("Async", req.async);
+			wait(auditStorageCore(self, req));
+			break;
 		} catch (Error& e) {
 			TraceEvent(SevInfo, "DDAuditStorageError", self->ddId)
 			    .errorUnsuppressed(e)
-			    .detail("AuditID", (audit == nullptr ? UID() : audit->id))
+			    .detail("AuditType", req.getType())
 			    .detail("Range", req.range)
-			    .detail("AuditType", req.type);
-			state Error err(e);
-			if (e.code() == error_code_audit_storage_error) {
-				auditState.setPhase(AuditPhase::Error);
-				wait(persistAuditState(self->txnProcessor->context(), auditState));
-			}
-			throw err;
-		}
-	} catch (Error& e) {
-		if (!req.async && !req.reply.isSet()) {
-			if (e.code() == error_code_audit_storage_error) {
-				req.reply.sendError(e);
+			    .detail("Async", req.async);
+			ASSERT(e.code() == error_code_actor_cancelled || e.code() == error_code_retry);
+			if (e.code() == error_code_retry) {
```

Comment:
Will we retry it indefinitely?

Reply:
The condition meets when `e.code() == error_code_retry`, captured by `wait(auditStorageCore(self, req));`. In `auditStorageCore`, only when `audit->retryCount <= SERVER_KNOBS->AUDIT_RETRY_COUNT_MAX` or `failed to read auditState` will do a retry. Let me add another guard there.